### PR TITLE
Updated calling codes as proposed at #114

### PR DESCRIPTION
**Pull request by herrjemand**
Changed the `callingCode` property to `callingCodes`.
Updated calling codes as proposed @nasht00 at #114.
Added Sardinian language as recognized minority language of Italy. Proposed by @chrispecoraro.

### CODE REVIEW

**Comment by petert82** [comment]
I hate to say this, but isn't the `callingCodes` field still wrong? :(

I mean, if it is supposed to be the international dialing code for the country, then all North American Numbering Plan countries should just have a `callingCode` of `"1"` (like Canada already has), but a lot of these countries have stuff like this:

```js
// Cayman Islands
"callingCodes": ["+1 345"]

// Dominican Republic
"callingCodes": ["+1 809", "+1 829", "+1 849"]
```

These seem to be a combination of the IDC and _every possible area code_ in that country. It seems a bit inconsistent to be doing this for some countries and not others.

What I had taken from the discussion on #114 was that every country should:

- Have one `callingCode`
- That code should just be the international dialling code for that country, with no local part

**Comment by herrniemand** [comment]
@petert82 
I renamed it to `callingCodes`, because some countries still have multiple calling codes(Dominican Republic, Netherlands). 
If we will make 35+ countries with calling code `+1` it will be even more pointless then it is now. http://www.nationsonline.org/maps/World-Calling-Codes-Map.jpg).

**Comment by herrniemand** [comment]
@petert82 besides. I tried to find expalanaition why there is three calling codes for Dominicans, and I failed. But every single website did said that there is three of them. http://fasteri.com/list/country-code-dialing-code-currency-timezone-of-countries.html

**Comment by petert82** [comment]
Why would it be pointless to have countries sharing a code? That's essentially what the NANP is. Surely that would just be the data reflecting reality?

I'd tend to defer to libphonenumber for determining this stuff, and it says that the country code for the Dominican Republic [is 1](http://giggsey.com/libphonenumber/?phonenumber=%2B1+809-544-4343&country=Gb&language=&region=).

I suspect the reason these sites are saying that DO has three codes is that there are only 3 area codes in the country, and so all numbers in the country _will_ start with one of those codes, but - unless I'm missing something - the 809, 829 and 849 parts of the number are _local_ area codes (see how 809 is part of the national number in the example I linked), they are not part of the country's international calling code.

**Comment by herrniemand** [comment]
@petert82 It would be useless, because if I am getting call from +7 I will think it is Russia when I will be getting call from Kazakhstan. Top level calling codes are not `tld`s. They represent geographical region, not a single country. What the point to know that you getting called from America, if there 35 countries with the same top level calling code?

**Comment by herrniemand** [comment]
@petert82 read that https://en.wikipedia.org/wiki/Telephone_numbers_in_the_Dominican_Republic.

**Comment by petert82** [comment]
I read that wiki page earlier, I think it supports what I already said: 809, 829 and 849 are area codes used in the Dominica Republic. They are not part of the international calling code.

If you want use the `callingCode` field to determine precisely where a number is coming from, then yes, I'd agree the data as it stands is not much use (for example; as is, you can't distinguish between USA & CAN), but I'm not sure a general purpose dataset like this is the best place to be trying to do that. To get close to that we'd need to include every area code for every country. This seems like a use case that would be much better suited to a dedicated phone number handling library.

As it stands, it's not clear to me what the `callingCode(s)` field is trying to be. Is it the telephone country code for that country, (as it is now for, say, the UK) or is it an enumeration of country code plus every possible area code in that country (as it _is_ for DO, but _isn't_ for, say, US)?

I'd argue it should be the former, because the latter doesn't seem practical, and the current inconsistency definitely does make the field useless.

**Comment by herrniemand** [comment]
@petert82 actually the whole IDD is fucked up. Seriously.

Zone +1
USA:
Айдахо (Idaho) — 208
Айова (Iowa) — 319, 515, 563, 641, 712
Алабама (Alabama) — 205, 251, 256, 334, 938
Аляска (Alaska) — 907
Аризона (Arizona) — 480, 520, 665, 623, 928
Арканзас (Arkansas) — 479, 501, 870
Вайоминг (Wyoming) — 307
Вашингтон (Washington) — 206, 253, 360, 425, 509
Вермонт (Vermont) — 802
Виргиния (Virginia) — 276, 434, 540, 571, 703, 757, 804
Висконсин (Wisconsin) — 262, 414, 534, 608, 715, 920
Гавайи (Hawaii) — 808
Делавэр (Delaware) — 302
Джорджия (Georgia) — 229, 404, 478, 678, 706, 762, 770, 912
Западная Виргиния (West Virginia) — 304, 681
Иллинойс (Illinois) — 217, 224, 309, 312, 331, 618, 630, 708, 773, 779, 815, 847, 872
Индиана (Indiana) — 219, 260, 317, 574, 765, 812
Калифорния (California) — 209, 213, 310, 323, 408, 415, 424, 442, 510, 530, 559, 562, 619, 626, 650, 657, 661, 707, 714, 747, 760, 805, 818, 831, 858, 909, 916, 925, 949, 951

WTF???

**Comment by herrniemand** [comment]
@petert82 at that point I am starting to ask my self "Why the hell would you make such a screwed up system?"

**Comment by nasht00** [comment]
I think the best solution is to have 2 columns:

- One showing the main digit only (for example, +1 for all the North American group), not an array, so that it works fine in dropdowns.
- One with the array of weird numbers, for those who want to use the database for number detection (scenario from @herrniemand about +7).

**Comment by herrniemand** [comment]
@nasht00 I think more logic would be either leave it as it is, or remove field.

**Comment by nasht00** [comment]
@herrniemand I don't understand.
- Leave it as is: doesn't solve the issue at hand at all, which is the case of a dropdown to select the country and show the associated country code
- Remove field. What field? the calling code?? the entire reason for the database?

What I am saying is simple:
If some people need the main digit only (me, @petert82 ...), they use the column called callingCode (with only the main digit).
If some want the whole thing (for geo-detection), they use the column with the full array with local digits.

**Comment by herrniemand** [comment]
@nasht00 @petert82 
```json
"CallingCode":{
	"top": "+1",
	"sub": [
		"853",
		"534",
		"811",
	]
}
```

Like this?

**Comment by petert82** [comment]
Something like that would seem sensible to me, yeah. Though I'd go with something like `international` and `areaCodes` (or `local`) for the names. The `+` in the international code is probably extraneous, too.

**Comment by herrniemand** [comment]
@petert82 @nasht00 I have suggestion.
```json
"IDD":{
	"root": "+1",
	"suffixes": ["853", "534", "811"]
}
```
IDD - [International direct dialing](https://en.wikipedia.org/wiki/International_direct_dialing)

And I would keep `+` for `nice look` reasons.

**Comment by nasht00** [comment]
I approve

**Comment by herrniemand** [comment]
@mledoze what ya think mate? *)

**Comment by mledoze** [comment]
I also approve the format that you suggested. It'll make the calling codes more useful as they are now. 
Although, are you sure that `["853", "534", "811"]` are suffixes and not prefixes?

Thank you all for your contributions on this topic, I'm sorry to be come late, I was busy these days.

**Comment by herrniemand** [comment]
@mledoze sweet!

**Comment by herrniemand** [comment]
@mledoze sorry. Currently a bit busy. Hopefully will finish for the next week.

**Comment by mledoze** [comment]
@herrniemand no problem, take your time.

**Comment by herrniemand** [comment]
@mledoze *)

2015-03-13 22:07 GMT+13:00 Mohammed Le Doze <notifications@github.com>:

**Comment by herrniemand** [comment]
Mmm... LOVE IDD...
```js
   ],
    "United States": [
        "201",
        "202",
        "203",
        "205",
        "206",
        "207",
        "208",
        "209",
        "210",
        "212",
        "213",
        "214",
        "215",
        "216",
        "217",
        "218",
        "219",
        "220",
        "224",
        "225",
        "227",
        "228",
        "229",
        "231",
        "234",
        "239",
        "240",
        "248",
        "251",
        "252",
        "253",
        "254",
        "256",
        "260",
        "262",
        "267",
        "269",
        "270",
        "272",
        "274",
        "276",
        "281",
        "283",
        "301",
        "302",
        "303",
        "304",
        "305",
        "307",
        "308",
        "309",
        "310",
        "312",
        "313",
        "314",
        "315",
        "316",
        "317",
        "318",
        "319",
        "320",
        "321",
        "323",
        "325",
        "327",
        "330",
        "331",
        "334",
        "336",
        "337",
        "339",
        "346",
        "347",
        "351",
        "352",
        "360",
        "361",
        "364",
        "380",
        "385",
        "386",
        "401",
        "402",
        "404",
        "405",
        "406",
        "407",
        "408",
        "409",
        "410",
        "412",
        "413",
        "414",
        "415",
        "417",
        "419",
        "423",
        "424",
        "425",
        "430",
        "432",
        "434",
        "435",
        "440",
        "442",
        "443",
        "447",
        "458",
        "463",
        "464",
        "469",
        "470",
        "475",
        "478",
        "479",
        "480",
        "484",
        "501",
        "502",
        "503",
        "504",
        "505",
        "507",
        "508",
        "509",
        "510",
        "512",
        "513",
        "515",
        "516",
        "517",
        "518",
        "520",
        "530",
        "531",
        "534",
        "539",
        "540",
        "541",
        "551",
        "559",
        "561",
        "562",
        "563",
        "564",
        "567",
        "570",
        "571",
        "573",
        "574",
        "575",
        "580",
        "585",
        "586",
        "601",
        "602",
        "603",
        "605",
        "606",
        "607",
        "608",
        "609",
        "610",
        "612",
        "614",
        "615",
        "616",
        "617",
        "618",
        "619",
        "620",
        "623",
        "626",
        "628",
        "629",
        "630",
        "631",
        "636",
        "641",
        "646",
        "650",
        "651",
        "657",
        "660",
        "661",
        "662",
        "667",
        "669",
        "678",
        "681",
        "682",
        "701",
        "702",
        "703",
        "704",
        "706",
        "707",
        "708",
        "712",
        "713",
        "714",
        "715",
        "716",
        "717",
        "718",
        "719",
        "720",
        "724",
        "725",
        "727",
        "730",
        "731",
        "732",
        "734",
        "737",
        "740",
        "743",
        "747",
        "754",
        "757",
        "760",
        "762",
        "763",
        "765",
        "769",
        "770",
        "772",
        "773",
        "774",
        "775",
        "779",
        "781",
        "785",
        "786",
        "801",
        "802",
        "803",
        "804",
        "805",
        "806",
        "808",
        "810",
        "812",
        "813",
        "814",
        "815",
        "816",
        "817",
        "818",
        "828",
        "830",
        "831",
        "832",
        "843",
        "845",
        "847",
        "848",
        "850",
        "854",
        "856",
        "857",
        "858",
        "859",
        "860",
        "862",
        "863",
        "864",
        "865",
        "870",
        "872",
        "878",
        "901",
        "903",
        "904",
        "906",
        "907",
        "908",
        "909",
        "910",
        "912",
        "913",
        "914",
        "915",
        "916",
        "917",
        "918",
        "919",
        "920",
        "925",
        "928",
        "929",
        "930",
        "931",
        "934",
        "936",
        "937",
        "938",
        "940",
        "941",
        "947",
        "949",
        "951",
        "952",
        "954",
        "956",
        "959",
        "970",
        "971",
        "972",
        "973",
        "975",
        "978",
        "979",
        "980",
        "984",
        "985",
        "989"
    ],
```

**Comment by herrniemand** [comment]
@mledoze FINISHED!!! :beers:

**Comment by herrniemand** [comment]
Just to explain Antarctica: It does not have calling code, because each station, depending on its nationality, has its own calling code.

**Comment by herrniemand** [comment]
@mledoze P.S. You need to regenerate `gh-pages` *)

**Comment by andreineculau** [comment]
FWIW, IDD (aka exit code) is being confused with the calling code in this PR

**Comment by herrniemand** [comment]
@andreineculau "Country calling codes or country dial in codes are telephone dialing prefixes for the member countries of the International Telecommunication Union (ITU). They are defined by the ITU-T in standards E.123 and E.164. The prefixes enable international direct dialing (IDD), and are also referred to as international subscriber dialing (ISD) codes." https://en.wikipedia.org/wiki/List_of_country_calling_codes

**Comment by andreineculau** [comment]
Correct. "country calling codes [..] enable international direct dialing", same way area codes enable national direct dialing.

Note: "prefixes" on the wiki page refers to IDD not ISD (=country calling codes). For a simple but clear disambiguation see the nice list here https://www.spotthelost.com/isd-codes.php

TL;DR IDD is enabled by/made of: IDD codes (hidden from most/all?! users by the + sign), ISD codes and area/mobile operator codes prefixing the local number

**Comment by herrniemand** [comment]
@andreineculau Oh. Ok. I see what you mean. This will be hard one *)
https://en.wikipedia.org/wiki/List_of_international_call_prefixes

**Comment by herrjemand** [comment]
...one day I will finish this pull... but for now :beers:

**Comment by mledoze** [comment]
Hello from the past :)
@herrjemand what do we do about this pull request?

**Comment by compojoom** [comment]
@herrjemand - are you able to update your pull request? It would be awesome if we could have this in the library.

**Comment by ialokim** [comment]
I've been building a [json file](https://github.com/ialokim/android-phone-field/blob/master/raw/countries-generator/countries.json) containing all calling codes as stated on Wikipedia for my [android-phone-field](https://github.com/ialokim/android-phone-field) project.

It has the following format:
```json
{
    "name": "Cape Verde (Kabu Verdi)",
    "iso2": "cv",
    "dialCode": "238",
    "priority": 1,
    "areaCodes": null
  },
  {
    "name": "Caribbean Netherlands",
    "iso2": "bq",
    "dialCode": "599",
    "priority": 1,
    "areaCodes": [
      "3",
      "4",
      "7"
    ]
},
```

I think this information could be quite easily added to this project. Not sure though when I find the time to look into this, so if someone is willing to work on this, please go ahead!

**Comment by mledoze** [comment]
Hello all, I just created the pull request #326 to merge the work of @herrjemand; as I explained there, since I'm not able to update the `master` branch of his fork, I created a new pull request. I will close this one for now.
